shared/src/main/scala/pl/metastack/metarouter/Route.scala

```diff
@@ -1,224 +1,17 @@
-    , tr  : Fill   .Aux[TLen, HR, TR]             // Infers TR
-
-    , hc  : Case1 .Aux[f.type, H, HR]              // Maps head
-    , mt  : Mapper.Aux[f.type, ROUTE, HR :: TR]   // Maps tail
-    , trav: ToTraversable.Aux[HR :: TR, List, HR]  // Converts HList to List
-  ): HR =
-    r.path.map(f).toList[HR].combineAll
-}
-
-trait RouteBase[ROUTE <: HList] {
-  def path: ROUTE
-
-  def fold[
-    H, HR: Monoid
-  , T <: HList, TR <: HList
-  , TLen <: Nat
-  ](f: Id ~>> HR)(implicit
-      cons: IsHCons.Aux[ROUTE, H, T]
-    , tlen: Length .Aux[T, TLen]
-    , tr  : Fill   .Aux[TLen, HR, TR]
-
-    , hc  : Case1  .Aux[f.type, H, HR]
-    , mt  : Mapper .Aux[f.type, ROUTE, HR :: TR]
-    , trav: ToTraversable.Aux[HR :: TR, List, HR]
-  ): HR = Route.fold(this, f)
 }
 
-case class Route[ROUTE <: HList] private (pathElements: ROUTE) extends RouteBase[ROUTE] {
-  def as[T]: MappedRoute[ROUTE, T] = MappedRoute[ROUTE, T](this)
-
-  def path = pathElements
-
-  def fill()(implicit map: FlatMapper.Aux[Route.ConvertArgs.type, ROUTE, HNil]):
-    InstantiatedRoute[ROUTE, HNil] =
-      InstantiatedRoute[ROUTE, HNil](this, HNil)
-
-  def fill[T, Params <: HList, ParamsCount <: Nat](arg: T)
-    (implicit
-      ev: FlatMapper.Aux[Route.ConvertArgs.type, ROUTE, Params],
-      ev2: Length.Aux[Params, ParamsCount],
-      ev3: ParamsCount =:= Nat._1,
-      ev4: Params =:= (T :: HNil)
-    ): InstantiatedRoute[ROUTE, T :: HNil] =
-      InstantiatedRoute[ROUTE, T :: HNil](this, arg :: HNil)
-
-  def fillN[L <: HList, TP <: Product, Params <: HList, ParamsCount <: Nat](args: TP)
-    (implicit
-      ev: FlatMapper.Aux[Route.ConvertArgs.type, ROUTE, Params],
-      ev2: Length.Aux[Params, ParamsCount],
-      ev3: ParamsCount <:!< Nat._1,
-      hl: Generic.Aux[TP, L],
-      ev4: Params =:= L
-    ): InstantiatedRoute[ROUTE, L] =
-      InstantiatedRoute[ROUTE, L](this, hl.to(args))
-
-  def parse[L <: HList](s: String)
-    (implicit map: FlatMapper.Aux[Route.ConvertArgs.type, ROUTE, L]):
-    Option[InstantiatedRoute[ROUTE, L]] =
-  {
-    import shapeless.HList.ListCompat._
-
-    def m[R <: HList](r: R, s: Seq[String]): Option[HList] =
-      (r, s) match {
-        case (HNil, Nil) => Some(HNil)
-        case (_,    Nil) => None
-        case (HNil, _)   => None
-        case ((rH: String)  #: rT, sH :: sT) if rH == sH => m(rT, sT)
-        case ((rH: String)  #: rT, sH :: sT) => None
-        case ((arg: Arg[_]) #: rT, sH :: sT) =>
-          arg.parseableArg.urlDecode(sH).flatMap { decoded =>
-            m(rT, sT).map(decoded :: _)
-          }
-      }
-
-    m[ROUTE](pathElements, Route.split(s)).map { x =>
-      // Using asInstanceOf as a band aid since compiler isn't able to confirm the type.
-      InstantiatedRoute(this, x.asInstanceOf[L])
-    }
-  }
-
+case class Route[ROUTE <: HList] private[metarouter] (pathElements: ROUTE) {
```

Comment:
Good point! I guess, we shouldn't be to prescriptive. What do you think about the following helper functions/classes that are internal and I wouldn't consider part of the public API:

https://github.com/MetaStack-pl/MetaRouter/blob/apply/shared/src/main/scala/pl/metastack/metarouter/Router.scala#L26
https://github.com/MetaStack-pl/MetaRouter/blob/apply/shared/src/main/scala/pl/metastack/metarouter/Router.scala#L36